### PR TITLE
fix(dia.Linkview): use defaultLabel.size for size calculations

### PR DESCRIPTION
**Pull request by zbynekstara**
## Description

Make it possible to specify `defaultLabel.size` when initializing a Link and have its properties merged with the properties of `label.size`.

## Motivation and Context

Currently, `markup`, `attrs`, and `position` arguments of labels get mixed in with `defaultLabel` arguments:
- `defaultLabel.markup` + `label.markup` (in `dia.LinkView.updateLabels()`)
- `defaultLabel.attrs` + `label.attrs` (in `dia.LinkView.updateLabels()`)
- `defaultLabel.position` + `label.position` (in `dia.LinkView.updateLabelPositions()`)

The same logic should apply to `size` argument of labels:
- `defaultLabel.size` + `label.size` (in `dia.LinkView.updateLabels()`)

## Notes

TypeScript definition for `size` already exists.
✅ Source code changed.
✅ Test provided.
✅ Docs changed (`defaultLabel` documentation added where appropriate).
✅ Tutorials changed (`defaultLabel` documentation added where appropriate).

### CODE REVIEW

File: docs/src/joint/api/dia/Link/intro.html [hunk 1]
```diff
@@ -44,14 +44,11 @@ <h4 id="dia.Link.intro.presentation">Presentation</h4>
 
 <p>As we can see, the <code>joint.shapes.standard.Link</code> shape consists of two <q>subelements</q>: one SVGPathElement named <code>'wrapper'</code> and one SVGPathElement named <code>'line'</code>. The <a href="#dia.Link.intro.styling"><code>attrs</code></a> object refers to the subelements' names (<q>selectors</q>) to provide SVG attributes to these constituent SVGElements.</p>
 
+<p><div class="docs-important-note">Direct use of <code>joint.dia.Link</code> is deprecated. Use links from <a href="#shapes.standard"><code>standard</code></a> namespace.</div></p>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/intro.html:47]: "Use links from **the**..."

File: docs/src/joint/api/dia/Link/intro.html [hunk 2]
```diff
@@ -60,24 +57,23 @@ <h4 id="dia.Link.intro.presentation">Presentation</h4>
 &lt;g class=&quot;marker-arrowheads&quot;/&gt;
 &lt;g class=&quot;link-tools&quot; /&gt;</code></pre>
 
-    <p>As you can see, the link consists of a couple of SVG path elements and couple of SVG group elements:</p>
+<p>As you can see, the link consists of a couple of SVG path elements and couple of SVG group elements:</p>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/intro.html:60]: "and **a** couple of SVG group ..."

File: docs/src/joint/api/dia/Link/intro.html [hunk 3]
```diff
@@ -60,24 +57,23 @@ <h4 id="dia.Link.intro.presentation">Presentation</h4>
...
+<p>The subelements of the link can be optionally styled through link attributes. For example:</p>
 
+<pre><code data-lang="javascript">link.attr({
+    '.connection': { stroke: 'blue' },
+    '.marker-source': { fill: 'red', d: 'M 10 0 L 0 5 L 10 10 z' },
+    '.marker-target': { fill: 'yellow', d: 'M 10 0 L 0 5 L 10 10 z' }
+});</code></pre>
 
-<p>The <code>markup</code> attribute is not the only presentation attribute that may be specified for a <code>joint.dia.Link</code>. However, all the following Link properties are deprecated and included only for backwards compatibility. Use <a href="#linkTools">link tools</a> instead:</p>
+<p><div class="docs-important-note">The <code>markup</code> attribute is not the only presentation attribute that may be specified for a <code>joint.dia.Link</code>. However, the following four Link properties are deprecated and included only for backwards compatibility. Use <a href="#linkTools">link tools</a> instead.</div></p>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/intro.html:76]: I think `div` should be the parent of the `p` tag here. Think the `div` will close the `p` tag and render incorrectly in some cases.

File: docs/src/joint/api/dia/Link/labels.html [hunk 4]
```diff
@@ -26,23 +30,82 @@
     }
 }</code></pre>
 
-<h4>Markup</h4>
+<h4 id="dia.Link.labels.defaultLabel">Default Label</h4>
+
+<p>The <code>markup</code>, <code>size</code>, <code>attrs</code> and <code>position</code> objects defined on individual labels may be supplemented by properties from one of a hierarchy of <code>defaultLabel</code> objects, each of them following the same schema as above.</p>
+
+<ul>
+    <li>The highest precedence is held by <code>defaultLabel</code> object provided when <a href="#dia.Link.intro.labels">creating a new link instance</a>.</li>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/labels.html:38]: "...is held by **the**..."

File: docs/src/joint/api/dia/Link/labels.html [hunk 5]
```diff
@@ -26,23 +30,82 @@
...
+    <li>The second-highest precedence is held by <code>defaultLabel</code> object provided when <a href="#dia.Link.intro.customLink">defining a custom Link type</a>.</li>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/labels.html:39]: "...is held by **the**..."

File: docs/src/joint/api/dia/Link/labels.html [hunk 6]
```diff
@@ -26,23 +30,82 @@
...
+<p>If a <code>defaultLabel</code> object is found in one of the above sources, its <code>markup</code>, <code>size</code>, <code>attrs</code> and <code>position</code> properties act as a template into which the individually-specified label properties are deep-merged as appropriate. In addition, the properties of the highest-precedence <code>defaultLabel</code> object may prove to be important on their own; for example, the <code>linkView.addLabel()</code> <a href="#dia.LinkView.prototype.addLabel">shortcut function</a> is only capable of adding a default label to links.</p>
+
+<h4 id="dia.Link.labels.markup">Markup</h4>
+
+<p>The <code>markup</code> property specifies the markup of the label. It can be provided in three ways:</p>
+
+<ul>
+    <li>As a parsable SVG (e.g. <code>'&lt;rect /&gt;&lt;text /&gt;'</code>).</li>
+    <li>As a JSON array (e.g.<code>[{ tagName: 'rect', selector: 'body' }, { tagName: 'text', selector: 'label' }]</code>).</li>
+    <li>Or as an <code>joint.util.svg</code> <a href="#util.svg">ES6 tag template</a>:</li>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/Link/labels.html:52]: "Or as **a**..."

File: src/dia/LinkView.mjs [hunk 7]
```diff
@@ -457,7 +457,7 @@ export const LinkView = CellView.extend({
     },
 
 
-    // merge default label attrs into label attrs
+    // merge default label attrs into label attrs (or use builtin default label attrs if neither is provided)
```
Comment (jamesgeorgewilliams) [inline src/dia/LinkView.mjs:460]: "...**built-in**..."

File: src/dia/LinkView.mjs [hunk 8]
```diff
@@ -479,6 +479,22 @@ export const LinkView = CellView.extend({
         return merge({}, builtinDefaultLabelAttrs, defaultLabelAttrs, labelAttrs);
     },
 
+    // merge default label size into label size (no builtin)
```
Comment (jamesgeorgewilliams) [inline src/dia/LinkView.mjs:482]: "...**built-in**..."

File: tutorials/custom-links.html [hunk 9]
```diff
@@ -327,29 +327,36 @@ <h3 id="default-label">Default Label</h3>
...
+                <li><code>defaultLabel.size</code> - sets the default dimensions of labels created on this Link subtype.
+                No default value.</li>
                 <li><code>defaultLabel.attrs</code> - sets the default attributes of label subelements on this Link
                 subtype.
                 Uses selectors defined in <code>markup</code>.
                 <a href="/docs/jointjs#dia.Link.intro.builtinDefaultAttributes" target="_blank">See default.</a></li>
                 <li><code>defaultLabel.position</code> - sets the default position along the line at which labels will
                 be added for instances of this Link subtype.
-                Default is <code>0.5</code> (midpoint of the connection path).</li>
+                <a href="/docs/jointjs#dia.Link.intro.builtinDefaultAttributes" target="_blank">Default</a> is
```
Comment (jamesgeorgewilliams) [inline tutorials/custom-links.html:358]: "**The**... Default..."

File: tutorials/link-labels.html [hunk 10]
```diff
@@ -289,8 +307,15 @@ <h3 id="label-rotation">Label Rotation</h3>
...
             <p>Of course, it is also possible to change the appearance of your labels.
-                Simply specify custom <code>markup</code> in JSON format and pass some <code>attrs</code>.
+                To specify <a href="/docs/jointjs#dia.Link.labels.markup" target="_blank">custom markup</a>, you may
+                provide a <code>markup</code> object in JSON format or as a <code>joint.util.svg</code>
+                <a href="/docs/jointjs#util.svg" target="_blank">ES6 tag template</a>.
                 As a bonus, you can define custom selectors to identify individual components of your label.
+                To specify <a href="/docs/jointjs#dia.Link.labels.styling" target="_blank">custom attributes</a>, you
+                may provide an <code>attrs</code> object with SVG attributes or
+                <a href="special-attributes.html">JointJS special attributes</a>.
+                Finally, to specify <a href="/docs/jointjs#dia.Link.labels.size" target="_blank">custom size</a> of the
```
Comment (jamesgeorgewilliams) [inline tutorials/link-labels.html:317]: "Finally, to specify **the**..."

File: tutorials/link-labels.html [hunk 11]
```diff
@@ -354,9 +383,28 @@ <h3 id="label-styling">Label Styling</h3>
...
+                (If we had specified neither <code>size</code> nor an individual <code>ref</code> attribute on
+                <code>attrs/body</code>, the <code>defaultLabel.size</code> from the definition of the Link subtype
+                would have been used for reference instead.
+                However, if there had been no <code>defaultLabel</code> object in the Link subtype definition, or if its
+                <code>defaultLabel</code> had had no <code>size</code> property, the calculations would have used
+                <code>0</code> as the reference width and height.)</p>
+
+            <p>Additionally, note that in our example, the <code>x</code> and <code>y</code> position of the
+                <code>body</code> subelement is determined by reference to the <code>label</code> via a
+                <code>calc()</code> operation, as is the case for <code>asteriskBody</code> subelement which calculates
```
Comment (jamesgeorgewilliams) [inline tutorials/link-labels.html:406]: "... as is the case for **the**..."

File: tutorials/link-labels.html (hunk 10 above)
Comment (zbynekstara) [reply tutorials/link-labels.html:317]: I think this is actually fine, considering context:
- "To specify custom markup, you may..."
- "To specify custom attributes, you may..."
- "Finally, to specify custom size of the label, you may..."

**Pull request merged**